### PR TITLE
Take into account "fallback to container" when determining where native-image gets executed

### PR DESCRIPTION
**Pull request by yrodiere**
Sometimes we need to do things differently based on whether the native build happens within a container or not.

Before this patch, we used to determine that based exclusively on explicit configuration.

After this patch, we correctly take into account that we sometimes need to "fall back" to containers even though the configuration didn't mention anything about native-image, simply because native-image isn't installed.

The previous behavior used to lead to at least one bug: when debugging GraalVM's JVM, we determined the address to bind to based on whether we're running in a container or not, and
[here we used to make the wrong choice](https://github.com/quarkusio/quarkus/blob/635a848bb8022e4ca6e945af7e62edc32a288588/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java#L780-L784), resulting in the JVM debug agent being inaccessible from the Docker host.

See also https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/quarkus.2Enative.2Edebug-build-process/near/345399379

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildContainerRunner.java [hunk 1]
```diff
@@ -30,14 +31,18 @@ public abstract class NativeImageBuildContainerRunner extends NativeImageBuildRu
     public NativeImageBuildContainerRunner(NativeConfig nativeConfig, Path outputDir) {
         this.nativeConfig = nativeConfig;
         containerRuntime = nativeConfig.containerRuntime.orElseGet(ContainerRuntimeUtil::detectContainerRuntime);
-        log.infof("Using %s to run the native image builder", containerRuntime.getExecutableName());
```
Comment (yrodiere) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildContainerRunner.java:33]: I moved this to the build step in order to preserve the same relative order of logs.

**Review by zakkak — APPROVED** [review]
LGTM, thanks @yrodiere

**Comment by yrodiere** [comment]
The remaining test failures are caused by flaky tests AFAICT (I've seen there [in this other PR](https://github.com/quarkusio/quarkus/pull/32323#issuecomment-1492835450) for example).

@zakkak the code changed substantially since your review, but the spirit is the same, so I'll merge by the end of the day unless you object. Summary of what changed:

> @zakkak I had to make the `isContainerBuild` information available earlier in the build for the Kafka extensions, so including the `outputDir`/`nativeImageName` in the build item has proven problematic, because that could lead to circular dependencies...
> 
> I pushed an update that introduces a factory for the native image build runners, so that I can build that factory early, while still building the runners (which need the `outputDir`/`nativeImageName` in their constructor) later.
> 
> I think it's relatively safer than my previous approach, so I guess that problem with the Kafka extensions was a good thing after all :)

**Comment by zakkak** [comment]
Functionally it still looks good to me but we can (and I believe we should better) avoid the factory.

Please see https://github.com/quarkusio/quarkus/compare/main...zakkak:quarkus:native-container-build-build-process-debug-v2 and if you agree squash my commit in yours and update the PR. Thanks.

**Comment by yrodiere** [comment]
> Please see [main...zakkak:quarkus:native-container-build-build-process-debug-v2](https://github.com/quarkusio/quarkus/compare/main...zakkak:quarkus:native-container-build-build-process-debug-v2) and if you agree squash my commit in yours and update the PR. Thanks.

Works for me, thanks. I had actually started with this approach but backpedalled once I saw the many methods I'd have to change... But I agree it's better than what I had to do to introduce factories.

I squashed your commit and moved a log ("Using %s to run the native image builder") to the "setup" method in order to preserve the relative order of logs.

Let's see if CI agrees :)

**Review by zakkak — APPROVED** [review]
LGTM

Since I am now an author as well... Could someone please do another review? cc @Karm @geoand

**Review by geoand — APPROVED** [review]
Very good idea!

**Comment by yrodiere** [comment]
@zakkak Rebased and updated the code added in https://github.com/quarkusio/quarkus/pull/32335/files#diff-bbea232d6b9141ad313658b2587f9ec7e23d90c93cd79994dd14086a769d52d0R147

**Pull request merged**